Clean-up docs and API for `Trie` module.

matthewhammer opened the pull request:
- List main operations as `put`, `get` and `iter`.
- Example of using `Trie` directly, without `TrieMap` OO wrapper API.
- adds `get` as a synonym for `find`, making it easier to "find", perhaps.
- adds `iter` (adapted from `TrieMap`)
- Editing pass over old implementation notes, moving them after usage notes.
- Test `Trie` independently of `TrieMap` module.

crusso commented on src/Trie.mo:
```diff
@@ -733,6 +743,41 @@ module {
+  /// Returns an `Iter` over the key-value entries of the trie.
+  ///
+  /// Each iterator gets a _persistent view_ of the mapping, independent of concurrent updates to the iterated map.
+  public func iter<K, V>(t : Trie<K, V>) : I.Iter<(K, V)> {
```
indentation is inconsistent with rest of file.

matthewhammer replied:
I agree it looks strange, but I'm using the emacs mode's choices consistently everywhere.

I think there are two things that make it look strangely (over) indented here:
 - `object` around the entire thing, to create the `Iter`
 - double `switch` statement, the "best" syntax for showing off the emacs mode's worst behavior.
 
 I can manually remove the double-indentation for the inner `switch`, but what else would you change?

matthewhammer replied:
Committed your format suggestions.  Thanks!

crusso commented on test/TrieTest.mo:
```diff
@@ -0,0 +1,38 @@
+import Trie "mo:base/Trie";
```
It would be nice to add alot more test here. The matchers library should be good enough. There's some examples in other tests (i.e. testText.mo)

matthewhammer replied:
I agree, but that sounds like a worthwhile effort for a follow-up PR that focuses on expanding the tests, which I agree are lacking.

Really, the "test" I have here is just to test the example that I provided in the doc; it's not meant to serve as what you are asking about now.  I will rename this file to TrieExamples to make that clearer and plan to do a follow up, using matchers and the other tests as examples, per your advice.

crusso replied:
Ok, that makes sense.

crusso commented on src/Trie.mo:
```diff
@@ -326,6 +352,10 @@ module {
+
   ///  find the given key's value in the trie, or return null if nonexistent
   public func find<K, V>(t : Trie<K, V>, k : Key<K>, k_eq : (K, K) -> Bool) : ?V =
     label profile_trie_find : (?V) {
```
It would be nice to get rid of these profiling blocks now....

matthewhammer replied:
Done.

crusso reviewed (changes requested):
See comments

matthewhammer commented:
@crusso Thanks for taking a look.

I completely agree that this module deserves a test suite.  Perhaps as another PR though?  (Focusing on docs here).

crusso commented on src/Trie.mo:
```diff
@@ -733,6 +743,41 @@ module {
+  /// Each iterator gets a _persistent view_ of the mapping, independent of concurrent updates to the iterated map.
+  public func iter<K, V>(t : Trie<K, V>) : I.Iter<(K, V)> {
+    object {
+      var stack = ?(t, null) : List.List<Trie<K, V>>;
+      public func next() : ?(K, V) {
+        switch stack {
+        case null { null };
+        case (?(trie, stack2)) {
+               switch trie {
+               case (#empty) {
+                      stack := stack2;
+                      next()
+                    };
+               case (#leaf({keyvals = null})) {
+                      stack := stack2;
+                      next()
+                    };
+               case (#leaf({size = c; keyvals = ?((k, v), kvs)})) {
+                      stack := ?(#leaf({size=c-1; keyvals=kvs}), stack2);
+                      ?(k.key, v)
+                    };
+               case (#branch(br)) {
+                      stack := ?(br.left, ?(br.right, stack2));
+                      next()
+                    };
+               }
+             }
```
```suggestion
      var stack = ?(t, null) : List.List<Trie<K, V>>;
      public func next() : ?(K, V) {
        switch stack {
          case null { null };
          case (?(trie, stack2)) {
              switch trie {
                 case (#empty) {
                     stack := stack2;
                     next()
                  };
                 case (#leaf({ keyvals = null })) {
                     stack := stack2;
                     next()
                 };
               case (#leaf({ size = c; keyvals = ?((k, v), kvs)})) {
                   stack := ?(#leaf({ size = c-1; keyvals = kvs}), stack2);
                   ?(k.key, v)
               };
               case (#branch(br)) {
                   stack := ?(br.left, ?(br.right, stack2));
                   next()
               };
             }
           }
```
something like that (not too sure about the record spacing - check what I did elsewhere I guess or consult the guidelines which I was trying to follow (modulo misunderstanding).

crusso replied:
(generally, indent by 2 spaces)

crusso commented:
Actually, looking at the code again there are quite a few helpers that seem to just be there to avoid instantiating type parameters.
Hopefully, Motoko should be able to infer all those now, letting your write the more efficient code you first wanted to write.
Perhaps try that in a separate commit (to see if it works). Using VSCode might be helpful since you'll see the errors right away (or Motoko mode in emacs if you have that working).

E.g. https://github.com/dfinity/motoko-base/blob/174f24d3f0a20d11ad91dd9d1bd91371f9f79186/src/Trie.mo#L526

matthewhammer commented:
@crusso PTAL

Updates:
 - removed profile labels
 - removed explicit type args
 - renamed test file to clarify it's just for examples (in the doc).

@crusso wrote
> Hopefully, Motoko should be able to infer all those now, letting your write the more efficient code you first wanted to write.

Indeed!  Motoko can now infer all of the explicit type arguments.  I've adjusted the code (simplifying it) accordingly.  Thanks for suggesting this, and the other review comments as well.

crusso commented on src/Trie.mo:
```diff
@@ -1,13 +1,78 @@
+/// This module provides an applicative (functional) hash map.
+/// Notably, each `put` produces a **new trie _and value being replaced, if any_**.
+///
+/// Those looking for more familar (imperative,
```
```suggestion
/// Those looking for more familiar (imperative,
```

crusso commented on src/Trie.mo:
````diff
@@ -1,13 +1,78 @@
+/// assert (n == ?42);
+/// ```
+///
+/// ## Implementation overview
+///
+/// A (hash) trie is a binary tree container for key-value pairs that
+/// consists of leaf and branch nodes.
+///
+/// Each internal **branch node**
+/// represents having distinguished its key-value pairs on a single bit of
+/// the keys.
+/// By following paths in the trie, we determine an increasingly smaller
+/// and smaller subset of the keys.
+///
+/// Each **leaf node** consists of an association list of key-value pairs.
+///
+/// Each non-empty trie node stores a size; we discuss that more below.
+///
+/// ### Adaptive depth
+///
+/// We say that a leaf is valid if it contains no more than `MAX_LEAF_SIZE`
+/// key-value pairs.  When a leaf node grows too large, the
+/// binary tree produces a new internal binary node, and splits the leaf into
+/// a pair of leaves using an additional bit of their keys' hash strings.
+///
+/// For small mappings, the trie structure consists of a single
+/// leaf, which contains up to MAX_LEAF_SIZE key-value pairs.
+///
+/// ### Cached sizes
 ///
+/// At each branch and leaf, we use a stored size to support a
+/// memory-efficient `toArray` function, which itself relies on
+/// per-element projection via `nth`; in turn, `nth` directly uses the
+/// O(1)-time function `size` for achieving an acceptable level of
+/// algorithmic efficiently.  Notably, leaves are generally lists of
+/// key-value pairs, and we do not store a size for each Cons cell in the
+/// list.
````
If you put the implementation notes in simple comments, they might not appear in the doc (is that too much detail for users who don't care about the implementation?)

matthewhammer replied:
changed this all to `//` as per the suggestion.

crusso commented on src/Trie.mo:
```diff
@@ -430,15 +435,15 @@ module {
           case (#leaf(l), _) {
             let (ll, lr) = splitAssocList<K, V>(l.keyvals, bitpos);
-            rec(bitpos, br(lf(ll), lf(lr)), tr)
+            rec(bitpos, branch(lf(ll), lf(lr)), tr)
```
Cool!

crusso commented on src/Trie.mo:
```diff
@@ -579,34 +567,34 @@ module {
+         leaf(AssocList.disj<Key<K>, V, W, X>(null, l.keyvals, key_eq, vbin), bitpos)
        };
-       case (#branch(b)) { br3(recR(b.left, bitpos + 1),recR(b.right, bitpos + 1)) };
+       case (#branch(b)) { branch(recR(b.left, bitpos + 1),
```
reformat this case?

matthewhammer replied:
Done.

crusso commented on src/Trie.mo:
```diff
@@ -579,34 +567,34 @@ module {
-         br3(rec(bitpos + 1, b1.left, b2.left),
-             rec(bitpos + 1, b1.right, b2.right))
+         branch(rec(bitpos + 1, b1.left, b2.left),
+                rec(bitpos + 1, b1.right, b2.right))
```
```suggestion
           rec(bitpos + 1, b1.right, b2.right))
```

crusso commented on src/Trie.mo:
```diff
@@ -621,46 +609,34 @@ module {
-          br3(rec(bitpos + 1, b1.left, b2.left),
-            rec(bitpos + 1, b1.right, b2.right))
+          branch(rec(bitpos + 1, b1.left, b2.left),
+                rec(bitpos + 1, b1.right, b2.right))
```
```suggestion
            rec(bitpos + 1, b1.right, b2.right))
```

crusso commented on src/Trie.mo:
```diff
@@ -733,6 +704,41 @@ module {
+      var stack = ?(t, null) : List.List<Trie<K, V>>;
+      public func next() : ?(K, V) {
+        switch stack {
+           case null { null };
+           case (?(trie, stack2)) {
+              switch trie {
+                 case (#empty) {
+                    stack := stack2;
+                    next()
+                 };
+                 case (#leaf({ keyvals = null })) {
+                    stack := stack2;
+                    next()
+                 };
+                 case (#leaf({ size = c; keyvals = ?((k, v), kvs) })) {
+                    stack := ?(#leaf({ size = c-1; keyvals = kvs }), stack2);
+                    ?(k.key, v)
+                 };
+                 case (#branch(br)) {
+                    stack := ?(br.left, ?(br.right, stack2));
+                    next()
+                 };
+               }
+            }
```
indent by 2 spaces. not three.

crusso replied:
(probably my fault from previous suggestion)

matthewhammer replied:
ah, I see -- no problem.

crusso commented on src/TrieMap.mo:
```diff
@@ -1,10 +1,12 @@
-/// For now, this class does not permit a direct `clone` operation (neither does `HashMap`), but it does permit creating iterators via `iter()`.  The effect is similar: Each iterator costs `O(1)` to create, but represents a fixed view of the mapping that does not interfere with mutations (it will _not_ view subsequent insertions or mutations, if any).
+/// - Unlike HashMap, the internal representation uses a functional representation (via `Trie` module).
+///
+/// - This class does not permit a direct `clone` operation (neither does `HashMap`), but it does permit creating iterators via `iter()`.  Each iterator costs `O(1)` to create, but represents a fixed view of the mapping that does not interfere with mutations (it will _not_ view subsequent insertions or mutations, if any).
```
```suggestion
/// - This class does not permit a direct `clone` operation (neither does `HashMap`), but it does permit creating iterators via `iter()`.  Each iterator costs `O(1)` to create, but represents a fixed view of the mapping that does not interfere with mutations (it will _not_ reflect subsequent insertions or mutations, if any).
```

crusso reviewed (approved):
LTME.
Minor nits.
Maybe check the guidelines about whether to capitalize comments (though I don't much care, I think @rossberg might)

matthewhammer commented:
Thanks again @crusso.  I attended to the lower case sentence starts.  Not sure how those arose, but thanks for catching them.

PTAL.

crusso reviewed (approved):
LGTM
